### ISSUE TITLE
OutputLayer.getWeightGradient(): MCXENT performs softmax on wrong dimension

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#345]
OutputLayer.getWeightGradient() softmax operation, line 134:
```java
INDArray pYGivenX = Nd4j.getExecutioner().execAndReturn(Nd4j.getOpFactory().createTransform("softmax",preOut),0);
```
final argument in above function call (i.e., dimension) should be 1, not 0.

For example, on MLPBackpropIrisExample (3 outputs):
preOut has shape [batchSize, 3] hence softmax operation should be dimension 1, not 0.
Easiest way to confirm this is to run the above example with breakpoints in OutputLayer.getWeightGradient() function and inspect the pYGivenX variable.

**Comment by AlexDBlack** [comment]
Also: I'm pretty sure your loss function gradient is wrong.
If you've got a copy of the book Bishop: Neural Networks for Pattern Recognition (1995) handy, cross-entropy is covered on page 240 (eq.6.136). Or alternatively http://neuralnetworksanddeeplearning.com/chap3.html, eq. 82.

Basically (line 136) it should be ```java pYGivenX.sub(labels);``` not ```java labels.sub(pYGivenX);```.

**Comment by agibsonccc** [comment]
Open a separate issue for the labels.sub. y - x is used in most places, an example here:
http://deeplearning.stanford.edu/wiki/index.php/Logistic_Regression_Vectorization_Example
